### ISSUE TITLE
Usability test results

### ISSUE THREAD

**Issue by nfraprado** [keepassxreboot/keepassxc#5832]
# Background

I created and ran some simple UX tests on KeePassXC using the template created by @clarissalimab.
The original issue can be found here: https://github.com/clarissalimab/ux/issues/5

# Users

The test was run with 5 users:
* 1 with no previous password manager experience.
* 2 users of other password managers, but with no previous experience with KeePassXC.
* 2 KeePassXC users.

# Tests

## Objective
The tests were chosen to evaluate features or menus on KeePassXC that seemed less intuitive to use or to find.

## Tasks

1. Change the database password
2. Generate new random password
3. Use the Auto-Type feature
4. Check leaked passwords with Have I Been Pwned

## Scenarios

1. You accidentally exposed the current password for your database, and to prevent possible leaks you need to change it. **Expected behavior: "Database" > "Database settings" > "Security" > "Change Password", enter the new password twice and "OK".**
2. You want to create a new account on a website, and to make it very secure, you need to generate a new random password for it. **Expected behavior: Click on the plus icon (or "Entries" > "New Entry...") then on the Password field, click on the dice icon and click on "Apply Pasword" and "OK".**
3. You were creating a new account on a website and you noticed that it doesn't allow you to paste the password from the password manager into the field, so you need the password manager to type it in for you. **Expected behavior: Select password entry and click the keyboard icon (or "Entries" > "Perform Auto-Type")**
**Note: The user should be instructed to open a second window with a browser or terminal to act as the website being inputted into**
4. You learned that an online service you use suffered a big password leak, and since you learned from a friend that KeePassXC can give you this information, you need to check if one of your passwords were part of the leak. **Expected behavior: "Database" > "Database Reports..." > "HIBP" > "Perform Online Analysis".**

# Results

## Task 1

*4 of 5 participants succeeded.*

**What went well?**

After the users reached the "Database Settings" screen, they easily found where and how to change the database password.

**What were the challenges?**

Almost all users were expecting to find it under the "Settings" menu displayed as a Gear icon on the top bar, but that shows global application settings instead.

On top of that, opening that menu makes the options on the "Database" drop-down menu, which contains the "Database Settings" menu, unavailable. That made it a lot harder for users to reach the right menu after entering this one, since then they first needed to understand that they needed to leave it and only then go to the right one.

Additionally, there was a bit of difficulty in finding the "OK" button, as it was all the way at the bottom of the screen, and users expected to find it where the "Cancel" button right below the password fields was.

**Conclusions**

The Gear icon should also lead to the database settings.

The application settings shouldn't disable the "Database" drop-down menu.

## Task 2

*5 of 5 participants succeeded.*

**What went well?**

As soon as the users reached the "New Entry" screen they found it easy to create a new entry, including clicking the little die icon to generate a random password.

**What were the challenges?**

A few users clicked on the Die icon on the top bar to generate the random password, but that screen doesn't allow the user to save the password to a new entry, only to copy it to the clipboard and to exit the screen. The users that entered this screen would generate the password, copy it to the clipboard, notice there wasn't a way to save it, then exit and go to the "New Entry" menu, where they would finally discover the little die icon on the password field to generate a random password for the new entry.

**Conclusions**

The "Password Generator" screen under the die icon should have a button to save the generated password to a new entry.

## Task 3

*4 of 5 participants succeeded.*

**What went well?**

The Keyboard icon on the top bar was very intuitive for the Auto-Type function. It was also easy to find this feature right-clicking an entry.

**What were the challenges?**

The users generally weren't expecting the program to automatically change to the last focused window, and would try to focus the field on the browser themselves after starting Auto-Type.

**Conclusions**

None.

## Task 4

*4 of 5 participants succeeded.*

**What went well?**

After the users finally reached the "HIBP" screen, which shows a "CAUTION" note explaining the risks of the feature and asks for confirmation from the user, they understood they had found the feature.

**What were the challenges?**

The users weren't expecting the program to have this feature at all. After explaining that it was present, the feature was always searched for first under the Tools drop-down menu.

The users took a lot of time to click on "Database Reports", as it seemed something more like passive statistics for the database. In this screen, the name "HIBP" wasn't informative, but users tried clicking on it as there were only 3 options.
The users only understood they were on the right screen after reading the "CAUTION" note.

Additionally, users had a bit of difficulty exiting this screen, as the "Close" button was all the way down at the bottom of the screen. One user said he was expecting a "Home" button of some sort to go back.

**Conclusions**

The "HIBP" screen should be moved under the "Tools" drop-down menu, as well as be renamed to something more immediately recognizable not only for those who are familiar with the "Have I Been Pwned" service, something generic like "Password leaks".

**Comment by droidmonkey** [comment]
This is some high caliber stuff, thank you!

**Comment by droidmonkey** [comment]
Here are some ideas, from a developers perspective, to address the challenges:

1. Database Settings
   * Add database settings to right click menu on the root group
   * Add a toolbar icon to the left of the application settings gear to access the current databases' settings
   * Show a list of open databases in the application settings with their status and some meta data. Offer buttons to jump directly to settings and security features.

2. Create New Entry
   * Agree the "standalone password generator" should offer the ability to create a new entry with the generated password into a selected database/group.
 
3. Auto-Type
   * This toolbar button was improved in recent development and now includes a warning to the user that typing is about to begin with the ability to cancel as well.

4. HIBP
   * I disagree with the recommendations, we consolidated all database reports in one location which makes them easy to navigate. These reports are database specific so the tools menu does not make sense.
   * We need to do a better job advertising this feature and add it to our documentation.

Did any of the users attempt to read the getting started or user guides in order to complete the tasks?

**Comment by nfraprado** [comment]
> Here are some ideas, from a developers perspective, to address the challenges:
> 
>     1. Database Settings
>        
>        * Add database settings to right click menu on the root group
>        * Add a toolbar icon to the left of the application settings gear to access the current databases' settings
>        * Show a list of open databases in the application settings with their status and some meta data. Offer buttons to jump directly to settings and security features.

The second one seems to be the one that would benefit the users the most, but all three would have positive impacts, from my observations.

>     2. Create New Entry
>        
>        * Agree the "standalone password generator" should offer the ability to create a new entry with the generated password into a selected database/group.

Great. In case the user didn't select any group (which would be the case for all users in my test), the entry should just be added to the default (root) group of the currently open database.

>     3. Auto-Type
>        
>        * This toolbar button was improved in recent development and now includes a warning to the user that typing is about to begin with the ability to cancel as well.

Nice. I didn't write any conclusions as I was a bit confused on how to proceed, but I think this would be a great step in the right direction.

>     4. HIBP
>        
>        * I disagree with the recommendations, we consolidated all database reports in one location which makes them easy to navigate. These reports are database specific so the tools menu does not make sense.
>        * We need to do a better job advertising this feature and add it to our documentation.

I see. It does make sense for it to be in a database specific setting. Still it seems to me to be a poorly advertised feature in the interface.

What if that menu was also exposed either in the top-down "Tools" menu or as another icon button, and considering the currently selected database (or all open databases)? Would that be possible from a developers perspective?
Maybe it could go to the "Health Check" screen, and not specifically to "HIBP".
Also, although "Database Reports" is shorter, it feels too generic to me, which is perhaps why users took a long time to click it. Perhaps "Database health and statistics" would be a more descriptive name for this new button.

Also, as previously noted, the "HIBP" name is of little information to those that do not know the service, and would be best advertised more generically as "Check password leaks" or something like that.

> Did any of the users attempt to read the getting started or user guides in order to complete the tasks?

Well, 2 of the users actually opened the help menu during test 4 (which was the most difficult), but didn't read it, either from intimidation or because they didn't think it was part of the test, so no. But the 4-task usability test took only about 15 minutes with each user, so I think we should take this as the user's first attempt in finding the feature, and once this fails, then they would search the manual.

Also, thank you for so promptly offer ways to address the challenges!

**Comment by phoerious** [comment]
IMHO we can even remove the global settings button and replace it with a database settings button. The latter one is arguably more important.

**Comment by tunbridgep** [comment]
1. Add an option to generate a new option OR update the master password from the standalone Password Generator.
2. A big heart icon could appear in the toolbar, with a heartbeat line going through it, which goes straight to the Health Check page.
3. The database settings page should have a button on it to go to general application settings, and vice versa

**Comment by kpapad904** [comment]
I have been using KeePassXC under Linux (KDE Neon and Ubuntu) for almost 3 years (thank you very much devs!) and previously KeePass for over a decade (first under Windows and in recent years under Linux desktop).

However while I appreciate KeePassXC's advanced features (particularly under Linux) I still can't get used to its UI/UX. I find the UI/UX of the old KeePass.exe (running under Mono) much more functional and intuitive (e.g. when previewing or editing an existing entry). So I still do all of my editing and housekeeping there.

I wonder if it is a Mono vs Qt issue ...

**Comment by tunbridgep** [comment]
KeePass is definitely more "professional" in it's design, but KeepassXC to me seems significantly easier to use, mostly because it lacks the features of KeePass that I won't use, which would otherwise be bloating up the menus (like triggers).

**Comment by lindhe** [comment]
I disagree that KeePass has a more professional  design. I just think it looks old to the point that it's hard for use for anyone used to more modern design patterns.

**Comment by tunbridgep** [comment]
"professional" and "old" are often synonymous :)

Really though the main difference is the organization of the menus. KeePass
has a lot of very filled menus. KeepassXC tends to do things as buttons.
That's totally fine and it's not a criticism, I just find it interesting
how both projects are designed
